components/ai_chat/content/browser/ai_chat_tab_helper.cc

```diff
@@ -300,20 +269,47 @@ void AIChatTabHelper::GetPageContent(GetPageContentCallback callback,
     // Ocred but `remaining_page_count_` was set to total pdf page count. Hence
     // status IDS_PDF_OCR_COMPLETED or IDS_PDF_OCR_NO_RESULT won't be set.
     CheckPDFA11yTree();
+    return;
+  }
+  if (base::Contains(kPrintPreviewRetrievalHosts, GetPageURL().host_piece()) &&
+      print_preview_extraction_delegate_ != nullptr) {
+    // Get content using a printing / OCR mechanism, instead of
+    // directly from the source.
+    print_preview_extraction_delegate_->Extract(
+        is_pdf,
+        base::BindOnce(&AIChatTabHelper::OnExtractPrintPreviewContentComplete,
+                       weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
   } else {
-    if (base::Contains(kPrintPreviewRetrievalHosts,
-                       GetPageURL().host_piece())) {
-      pending_get_page_content_callback_ = std::move(callback);
-      NotifyPrintPreviewRequested(false);
-    } else {
-      FetchPageContent(web_contents(), invalidation_token, std::move(callback));
-    }
+    page_content_fetcher_delegate_->FetchPageContent(
+        invalidation_token,
+        base::BindOnce(&AIChatTabHelper::OnFetchPageContentComplete,
+                       weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
   }
 }
 
-void AIChatTabHelper::PrintPreviewFallback(GetPageContentCallback callback) {
-  pending_get_page_content_callback_ = std::move(callback);
-  NotifyPrintPreviewRequested(IsPdf(web_contents()));
+void AIChatTabHelper::OnFetchPageContentComplete(
+    ConversationHandler::GetPageContentCallback callback,
+    std::string content,
+    bool is_video,
+    std::string invalidation_token) {
+  if (base::CollapseWhitespaceASCII(content, true).empty() && !is_video &&
+      print_preview_extraction_delegate_ != nullptr) {
+    DVLOG(1) << "Initiating print preview fallback";
+    print_preview_extraction_delegate_->Extract(
+        IsPdf(web_contents()),
+        base::BindOnce(&AIChatTabHelper::OnExtractPrintPreviewContentComplete,
+                       weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
+    return;
+  }
+  std::move(callback).Run(std::move(content), is_video,
+                          std::move(invalidation_token));
+}
+
+void AIChatTabHelper::OnExtractPrintPreviewContentComplete(
+    ConversationHandler::GetPageContentCallback callback,
+    std::string content) {
+  // Invalidation token not applicable for print preview OCR
```

Comment:
The third parameter of `GetPageContentCallback` is the `invalidation_token`, which helps not refetch adjacent content (e.g. video transcripts) if the underlying content didn't change between calls to `GetPageContent` (which is called every time the user submits a chat message). On some (most!) sites we want to re-fetch every time the content because it might have updated and it's just a simple local-only content extraction. However on  other sites, the content fetch results in network calls or might be expensive otherwise _and_ we know the content is static for now, so we use an invalidation token to say "use your cache for this content if the token is still the same as the one we extract cheaply".

Reply:
sweet, sgtm